posix/socket.c

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/sys/socket.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+
+static uint64_t _default_socket_devid = OE_DEVID_NONE;
+static oe_spinlock_t _default_socket_devid_lock = OE_SPINLOCK_INITIALIZER;
+
+void oe_set_default_socket_devid(uint64_t devid)
+{
+    oe_spin_lock(&_default_socket_devid_lock);
+    _default_socket_devid = devid;
+    oe_spin_unlock(&_default_socket_devid_lock);
+}
+
+uint64_t oe_get_default_socket_devid()
+{
+    oe_spin_lock(&_default_socket_devid_lock);
+    uint64_t ret = _default_socket_devid;
+    oe_spin_unlock(&_default_socket_devid_lock);
+    return ret;
+}
+
+int oe_socket_d(uint64_t devid, int domain, int type, int protocol)
+{
+    int ret = -1;
+    int sd;
+    oe_fd_t* sock = NULL;
+    oe_device_t* device;
+
+    if (devid == OE_DEVID_NONE)
+    {
+        /* Only one device today. */
+        devid = OE_DEVID_HOST_SOCKET_INTERFACE;
+    }
+
+    if (!(device = oe_device_table_get(devid, OE_DEVICE_TYPE_SOCKET_INTERFACE)))
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (!(sock = device->ops.socket.socket(device, domain, type, protocol)))
+    {
+        OE_RAISE_ERRNO_MSG(
+            oe_errno,
+            "devid=%ld domain=%d type=%d protocol=%d",
+            devid,
+            domain,
+            type,
+            protocol);
+        goto done;
+    }
+
+    if ((sd = oe_fdtable_assign(sock)) == -1)
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = sd;
+    sock = NULL;
+
+done:
+
+    if (sock)
+        sock->ops.fd.close(sock);
+
+    return ret;
+}
+
+int oe_socketpair(int domain, int type, int protocol, int retfd[2])
+{
+    int ret = -1;
+    ssize_t retval;
+    oe_fd_t* socks[2] = {0};
+    oe_device_t* device;
+    uint64_t devid = OE_DEVID_HOST_SOCKET_INTERFACE;
+
+    /* Resolve the device id. */
+    /* 2do: select based on address family */
```

Comment:
>2do [](start = 7, length = 3)

Nit: use standard spelling (i.e. TODO:) so that it shows up in global text searches

Reply:
Fixed. Thanks. I thought I got rid of all these.